File: nemonic_sdk_flutter/nemonic_sdk/ios/Classes/NemonicSdkController.swift
import Flutter
import NemonicSdk

class NemonicSdkController: NSObject {
    private var channel: FlutterMethodChannel
    private var printerController: NPrinterController!

    init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
        printerController = NPrinterController(self)
    }

    public func getDefaultConnectDelay() -> UInt32 {
        return printerController.getDefaultConnectDelay()
    }

    public func getConnectDelay() -> UInt32 {
        return printerController.getConnectDelay()
    }

    public func setConnectDelay(_ msec: UInt32) {
        return printerController.setConnectDelay(msec)
    }

    public func connect(_ printer: NPrinter, queueLabel: String? = nil) -> Int {
        return printerController.connect(printer, queueLabel: queueLabel)
    }

    public func disconnect() {
        printerController.disconnect()
    }

    public func getConnectState() -> Int {
        return printerController.getConnectState()
    }

    public func cancel() {
        printerController.cancel()
    }

    public func setPrintTimeout(_ enableAuto: Bool, _ manualTime: Int) {
        printerController.setPrintTimeout(enableAuto, manualTime)
    }

    public func print(_ printInfo: NPrintInfo) -> Int {
        return printerController.print(printInfo)
    }

    public func setTemplate(_ image: UIImage, withPrint: Bool, enableDither: Bool) -> Int {
        return printerController.setTemplate(image, withPrint: withPrint, enableDither: enableDither)
    }

    public func clearTemplate() -> Int {
        printerController.clearTemplate()
    }

    public func getPrinterStatus() -> Int {
        return printerController.getPrinterStatus()
    }

    public func getCartridgeType() -> Int {
        return printerController.getCartridgeType()
    }

    public func getPrinterName() -> NResultString {
        return printerController.getPrinterName()
    }

    public func getBatteryLevel() -> Int {
        return printerController.getBatteryLevel()
    }

    public func getBatteryStatus() -> Int {
        return printerController.getBatteryStatus()
    }
}

extension NemonicSdkController: NPrinterControllerDelegate {
    public func disconnected() {
        DispatchQueue.main.async {
            self.channel.invokeMethod("disconnected", arguments: nil)
        }
    }

    public func printProgress(index: Int, total: Int, result: Int) {
        let data: [String: Any] = [
            "index": index,
            "total": total,
            "result": result
        ]

        DispatchQueue.main.async {
            self.channel.invokeMethod("printProgress", arguments: data)
        }
    }

    public func printComplete(result: Int) {
        let data: [String: Any] = [
            "result": result
        ]

        DispatchQueue.main.async {
            self.channel.invokeMethod("printComplete", arguments: data)
        }
    }
}

File: nemonic_sdk_flutter/nemonic_sdk/android/src/main/java/com/mangoslab/nemonicsdk/nemonic_sdk/NemonicSdkController.java
package com.mangoslab.nemonicsdk.nemonic_sdk;

import android.content.Context;
import android.graphics.Bitmap;
import android.os.Handler;
import android.os.Looper;

import com.mangoslab.nemonicsdk.constants.NBatteryStatus;
import com.mangoslab.nemonicsdk.constants.NCartridgeType;
import com.mangoslab.nemonicsdk.constants.NConnectState;
import com.mangoslab.nemonicsdk.constants.NPrinterType;
import com.mangoslab.nemonicsdk.constants.NResult;
import com.mangoslab.nemonicsdk.constants.NResultString;
import com.mangoslab.nemonicsdk.INPrinterControllerCallback;
import com.mangoslab.nemonicsdk.NPrinter;
import com.mangoslab.nemonicsdk.NPrinterController;
import com.mangoslab.nemonicsdk.NPrintInfo;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class NemonicSdkController implements INPrinterControllerCallback {
    private static final String TAG = "NemonicSdkController";
    private MethodChannel channel;
    private NPrinterController printerController;

    public NemonicSdkController(MethodChannel channel, Context context) {
        this.channel = channel;
        printerController = new NPrinterController(context, this);
    }

    public int getDefaultConnectDelay() {
        return printerController.getDefaultConnectDelay();
    }

    public int getConnectDelay() {
        return printerController.getConnectDelay();
    }

    public void setConnectDelay(int msec) {
        printerController.setConnectDelay(msec);
    }

    public int connect(NPrinter printer) {
        return printerController.connect(printer);
    }

    public void disconnect() {
        printerController.disconnect();
    }

    public int getConnectState() {
        return printerController.getConnectState();
    }

    public void cancel() {
        printerController.cancel();
    }

    public void setPrintTimeout(boolean enableAuto, int manualTimeout) {
        printerController.setPrintTimeout(enableAuto, manualTimeout);
    }

    public int print(NPrintInfo printInfo) {
        return printerController.print(printInfo);
    }

    public int setTemplate(Bitmap image, boolean withPrint, boolean enableDither) {
        return printerController.setTemplate(image, withPrint, enableDither);
    }

    public int clearTemplate() {
        return printerController.clearTemplate();
    }

    public int getPrinterStatus() {
        return printerController.getPrinterStatus();
    }

    public int getCartridgeType() {
        return printerController.getCartridgeType();
    }

    public NResultString getPrinterName() {
        return printerController.getPrinterName();
    }

    public int getBatteryLevel() {
        return printerController.getBatteryLevel();
    }

    public int getBatteryStatus() {
        return printerController.getBatteryStatus();
    }

    @Override
    public void disconnected() {
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(new Runnable() {
            @Override
            public void run() {
                channel.invokeMethod("disconnected", null);
            }
        });
    }

    @Override
    public void printProgress(int index, int total, int result) {
        Map<String, Object> data = new HashMap<String, Object>();
        data.put("index", index);
        data.put("total", total);
        data.put("result", result);

        // channel.invokeMethod("printProgress", data);
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(new Runnable() {
            @Override
            public void run() {
                channel.invokeMethod("printProgress", data);
            }
        });
    }

    @Override
    public void printComplete(int result) {
        Map<String, Object> data = new HashMap<String, Object>();
        data.put("result", result);

        // channel.invokeMethod("printComplete", data);
        Handler handler = new Handler(Looper.getMainLooper());
        handler.post(new Runnable() {
            @Override
            public void run() {
                channel.invokeMethod("printComplete", data);
            }
        });
    }
}
